Modix.Services/Promotions/PromotionService.cs

```diff
@@ -1,72 +1,75 @@
-        private IPromotionRepository _repository;
+        private readonly Regex _badCharacterRegex = new Regex(@"[\u200B-\u200D\uFEFF]");
 
-        private SocketGuild CurrentGuild => _client.Guilds.First();
-        private IMessageChannel PromotionChannel => CurrentGuild.GetChannel(_promotionChannelId) as IMessageChannel;
+        private readonly DiscordSocketClient _client;
+        private readonly IPromotionRepository _repository;
+
+        private readonly ulong allowedToCommentRoleID;
+        private readonly ulong allowedToCreateRoleID;
+        private readonly ulong promotionChannelID;
 
-        public PromotionService(DiscordSocketClient client, IPromotionRepository repository)
+        public PromotionService(DiscordSocketClient client, IPromotionRepository repository, ModixConfig config)
         {
             _client = client;
             _repository = repository;
+            promotionChannelID = config.ChannelIdForPromotionCampaignAnnouncement;
+            allowedToCommentRoleID = config.RoleIdToAllowCommentingOnPromotionCampaign;
+            allowedToCreateRoleID = config.RoleIdToAllowCreatingPromotionCampaign;
         }
 
-        public Task<IEnumerable<PromotionCampaign>> GetCampaigns() => _repository.GetCampaigns();
-        public Task<PromotionCampaign> GetCampaign(int id) => _repository.GetCampaign(id);
+        private SocketGuild CurrentGuild => _client.Guilds.First();
+        private IMessageChannel PromotionChannel => CurrentGuild.GetChannel(promotionChannelID) as IMessageChannel;
+
+        public Task<IEnumerable<PromotionCampaign>> GetCampaigns()
+        {
+            return _repository.GetCampaigns();
+        }
+
+        public Task<PromotionCampaign> GetCampaign(int id)
+        {
+            return _repository.GetCampaign(id);
+        }
 
         public async Task ApproveCampaign(SocketGuildUser promoter, PromotionCampaign campaign)
         {
             ThrowIfNotStaff(promoter);
 
-            var foundUser = CurrentGuild.GetUser(campaign.UserId);
-            var foundRole = CurrentGuild.Roles.FirstOrDefault(d => d.Id == _regularRoleId);
+            var foundUser = CurrentGuild.GetUser(campaign.PromotionFor.DiscordUserId);
+            var foundRole = CurrentGuild.Roles.FirstOrDefault(d => d.Id == allowedToCommentRoleID);
 
             if (foundRole == null)
-            {
                 throw new InvalidOperationException("The server does not have a 'Regular' role to grant.");
-            }
 
             await foundUser.AddRoleAsync(foundRole);
 
             campaign.Status = CampaignStatus.Approved;
             await _repository.UpdateCampaign(campaign);
 
-            await PromotionChannel?.SendMessageAsync($"{MentionUtils.MentionUser(campaign.UserId)} has been promoted to Regular! 🎉");
+            if (PromotionChannel == null)
+                throw new NullReferenceException(nameof(PromotionChannel));
+            await PromotionChannel?.SendMessageAsync(
```

Comment:
Super minor, but if you're null-checking `PromotionChannel` then you can do `await PromotionChannel.SendMessageAsync` instead of `PromotionChannel?.SendMessageAsync`. It's problematic to write code like this anyway because I'm pretty sure trying to await `null` will cause an exception.